openvdb_ax/openvdb_ax/compiler/PointExecutable.h

```diff
@@ -0,0 +1,148 @@
+// Copyright Contributors to the OpenVDB Project
+// SPDX-License-Identifier: MPL-2.0
+
+/// @file compiler/PointExecutable.h
+///
+/// @authors Nick Avramoussis, Francisco Gochez, Richard Jones
+///
+/// @brief The PointExecutable, produced by the OpenVDB AX Compiler for
+///   execution over OpenVDB Points Grids.
+///
+
+#ifndef OPENVDB_AX_COMPILER_POINT_EXECUTABLE_HAS_BEEN_INCLUDED
+#define OPENVDB_AX_COMPILER_POINT_EXECUTABLE_HAS_BEEN_INCLUDED
+
+#include "../compiler/CustomData.h"
+#include "../compiler/AttributeRegistry.h"
+
+#include <openvdb/openvdb.h>
+#include <openvdb/points/PointDataGrid.h>
+
+#include <unordered_map>
+
+class TestPointExecutable;
+
+namespace llvm {
+class ExecutionEngine;
+class LLVMContext;
+}
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace ax {
+
+class Compiler;
+
+/// @brief Object that encapsulates compiled AX code which can be executed on a target point grid
+class PointExecutable
+{
+public:
+    using Ptr = std::shared_ptr<PointExecutable>;
+    ~PointExecutable();
+
+    /// @brief  Copy constructor. Shares the LLVM constructs but deep copies the
+    ///   settings. Multiple copies of an executor can be used at the same time
+    ///   safely.
+    PointExecutable(const PointExecutable& other);
+
+    ////////////////////////////////////////////////////////
+
+    /// @brief executes compiled AX code on target grid
+    void execute(points::PointDataGrid& grid) const;
+
+    ////////////////////////////////////////////////////////
+
+    /// @brief  Set a specific point group to execute over. The default is none,
+    ///   which corresponds to all points. Note that this can also be compiled
+    ///   into the AX function using the ingroup("mygroup") method.
+    /// @warning  If the group does not exist during execute, a runtime error
+    ///   will be thrown.
+    /// @param name  The name of the group to execute over
+    void setGroupExecution(const std::string& name);
+    /// @return  The points group to be processed. Default is empty, which is
+    ///   all points.
+    const std::string& getGroupExecution() const;
+
+    /// @brief  Set the behaviour when missing point attributes are accessed.
+    ///    Default behaviour is true, which creates them with default initial
+    ///    values. If false, a missing attribute runtime error will be thrown
+    ///    on missing accesses.
+    /// @param flag  Enables or disables the creation of missing attributes
+    void setCreateMissing(const bool flag);
+    /// @return  Whether this executable will generate new point attributes.
+    bool getCreateMissing() const;
+
+    /// @brief  Set the threading grain size. Default is 1. A value of 0 has the
+    ///   effect of disabling multi-threading.
+    /// @param grain The grain size
+    void setGrainSize(const size_t grain);
+    /// @return  The current grain size
+    size_t getGrainSize() const;
```

Comment:
I don't like the grain size being a setting of the executable. I think it should be an argument to the execute() method with a default value. I'd also like to see a threaded=true argument to align with other similar methods in VDB such as LeafManager::foreach().